pkg/scheduler/metrics/resources/resources.go

```diff
@@ -0,0 +1,237 @@
+	for _, p := range pods {
+		reqs, limits, terminal := podRequestsAndLimitsByLifecycle(p, reuseReqs, reuseLimits)
+		if terminal {
+			// terminal pods are excluded from resource usage calculations
+			continue
+		}
+		for _, t := range []struct {
+			desc  resourceLifecycleDescriptors
+			total v1.ResourceList
+		}{
+			{
+				desc:  podResourceDesc.requests,
+				total: reqs,
+			},
+			{
+				desc:  podResourceDesc.limits,
+				total: limits,
+			},
+		} {
+			for resourceName, val := range t.total {
+				var unitName string
+				switch resourceName {
+				case v1.ResourceCPU:
+					unitName = "cores"
+				case v1.ResourceMemory:
+					unitName = "bytes"
+				case v1.ResourceStorage:
+					unitName = "bytes"
+				case v1.ResourceEphemeralStorage:
+					unitName = "bytes"
+				default:
+					switch {
+					case v1helper.IsHugePageResourceName(resourceName):
+						unitName = "bytes"
+					case v1helper.IsAttachableVolumeResourceName(resourceName):
+						unitName = "integer"
+					}
+				}
+				var priority string
+				if p.Spec.Priority != nil {
+					priority = strconv.FormatInt(int64(*p.Spec.Priority), 10)
+				}
+				recordMetricWithUnit(ch, t.desc.total, p.Namespace, p.Name, p.Spec.NodeName, p.Spec.SchedulerName, priority, resourceName, unitName, val)
+			}
+		}
+	}
+}
+
+func recordMetricWithUnit(
+	ch chan<- metrics.Metric,
+	desc *metrics.Desc,
+	namespace, name, nodeName, schedulerName, priority string,
+	resourceName v1.ResourceName,
+	unit string,
+	val resource.Quantity,
+) {
+	if val.IsZero() {
+		return
+	}
+	ch <- metrics.NewLazyConstMetric(desc, metrics.GaugeValue,
+		val.AsApproximateFloat64(),
+		namespace, name, nodeName, schedulerName, priority, string(resourceName), unit,
+	)
+}
+
+// podRequestsAndLimitsByLifecycle returns a dictionary of all defined resources summed up for all
+// containers of the pod. If PodOverhead feature is enabled, pod overhead is added to the
+// total container resource requests and to the total container limits which have a
+// non-zero quantity. The caller may avoid allocations of resource lists by passing
+// a requests and limits list to the function, which will be cleared before use.
+// This method is the same as v1resource.PodRequestsAndLimits but avoids allocating in several
+// scenarios for efficiency.
+func podRequestsAndLimitsByLifecycle(pod *v1.Pod, reuseReqs, reuseLimits v1.ResourceList) (reqs, limits v1.ResourceList, terminal bool) {
+	switch {
+	case len(pod.Spec.NodeName) == 0:
+		// unscheduled pods cannot be terminal
+	case pod.Status.Phase == v1.PodSucceeded, pod.Status.Phase == v1.PodFailed:
+		terminal = true
+		// TODO: add a condition for deleted pods where all containers are terminated once scheduler
+		//   has the same logic
+	}
+	if terminal {
+		return
+	}
+
+	// attempt to reuse the maps if passed, or allocate otherwise
+	reqs, limits = reuseOrClearResourceList(reuseReqs), reuseOrClearResourceList(reuseLimits)
+
+	for _, container := range pod.Spec.Containers {
```

Comment:
Why not use `v1resource.PodRequestsAndLimits`?

sigh for yet another duplication I had forgotten about.

Reply:
Because that allocates every time, and changing the public signature is expensive, which is why we generally don't over consolidate things like that.  I am not averse to adding a new model that allows reuse though in this case.

Reply:
I see.
What if you add a new function in v1resource that accepts already allocated maps and then you have this function and `v1resource.PodRequestsAndLimits` call that?

Reply:
Yeah that's what I was considering.  Worried about misuse, but i'd be ok sharing the impl now that we no longer need to track lifecycle here.

Reply:
Please do